### PR TITLE
Fixes and additional sensor support for Si7006, SHT21, HTU21D sensors

### PR DESCRIPTION
**Pull request by xyzzy42**
This driver can support other sensors that use the same commands and conversion formulas.

Add SHT21 and HTU21D support.

The math to convert the values was rounding down to an integer.

See individual commits for much more detail.

### CODE REVIEW

**Comment by xyzzy42** [comment]
This allows the driver to support the three types, as described by @pabigot three years ago back in this comment, https://github.com/zephyrproject-rtos/zephyr/pull/22862#pullrequestreview-359397725

Prior work to try to add these sensors: #11458, #22862, #30185, #38318, #51269.

File: dts/bindings/sensor/silabs,si7006.yaml
Comment (fkokosinski) [inline dts/bindings/sensor/silabs,si7006.yaml:1]: Please don't remove old copyright entries.

File: drivers/sensor/si7006/si7006.c
Comment (fkokosinski) [inline drivers/sensor/si7006/si7006.c:7]: Could you extract the generic part of this driver, which would be used by the si7006 and the sht21 drivers? Having it all in one driver named `si7006` makes it confusing.

File: dts/bindings/sensor/silabs,si7006.yaml
Comment (xyzzy42) [reply dts/bindings/sensor/silabs,si7006.yaml:1]: It's a complete re-write of the file.

File: drivers/sensor/si7006/si7006.c
Comment (xyzzy42) [reply drivers/sensor/si7006/si7006.c:7]: There is just one driver.  The only difference to support both sensors is literally two bits in one byte.

The driver should probably be renamed sht21, since that's the original device.  I can do that in a followup PR.  I find it's much better to do mass-renames in a PR of their own, not combined with a PR that also changes the code in other ways.

Comment (MaureenHelm) [reply drivers/sensor/si7006/si7006.c:7]: You need to define `DT_DRV_COMPAT` somewhere for the DT_INST macros below to work. I'd expect to see something like this to support multiple compatibles in the driver: https://github.com/zephyrproject-rtos/zephyr/blob/600e7b7f21cfd179609372bb4f5e3bc5d33d19c2/drivers/sensor/tmp108/tmp108.c#L406-L414

File: dts/bindings/sensor/silabs,si7006.yaml
Comment (MaureenHelm) [reply dts/bindings/sensor/silabs,si7006.yaml:1]: Agree with @fkokosinski

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add the new compatible to `tests/drivers/build_all/sensor`

File: drivers/sensor/si7006/si7006.c
Comment (xyzzy42) [reply drivers/sensor/si7006/si7006.c:7]: I assure you, this driver works.

I don't use the DT_INST macros.  The non "INST" versions work fine.  This seems nicer than needing to undef and redefine DT_DRV_COMPAT between each call.

File: dts/bindings/sensor/silabs,si7006.yaml
Comment (xyzzy42) [reply dts/bindings/sensor/silabs,si7006.yaml:1]: Added back.

File: drivers/sensor/si7006/si7006.c
Comment (xyzzy42) [reply drivers/sensor/si7006/si7006.c:7]: I thought more about splitting the driver.  It doesn't make sense.  tmp108 + as6212 show an example of one driver for two compatible devices.  This is like that.

Splitting it would mean putting everything into the common file.  Then making various static and private data global, i.e. the api definition object, the type for the data and config structs, the init function.  Zephyr doesn't use LTO, so there is a cost to making things non-static.  Then shc21.c and si7006.c become just one line each:
```c
DT_FOREACH_STATUS_OKAY_VARGS(silabs_si7006, SI7006_DEFINE, SI7006_READ_OLD_TEMP);
```
And likewise for sht21.

It's apparent that this just a convoluted way to add more source files solely for the purpose of adding more source files.

And your original issue, that the code for different-but-compatible sensors is in a file that uses the name si7006, is not improved in any way.  All the code (less one line) is still in one file with one name.  

It would be reasonable to mass rename si7006 to something else.  I think sht21, being the original, would be more correct.  Or another name.

**Comment by xyzzy42** [comment]
> Please add the new compatible to `tests/drivers/build_all/sensor`

Added.  I put it at the end, so the i2c addresses would be used in order.  This seems more useful than putting it near the si7006 binding.

File: drivers/sensor/si7006/si7006.c
Comment (MaureenHelm) [reply drivers/sensor/si7006/si7006.c:7]: > I don't use the DT_INST macros. The non "INST" versions work fine. This seems nicer than needing to undef and redefine DT_DRV_COMPAT between each call.

Please keep the INST versions to be consistent with other drivers

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
>> Please add the new compatible to tests/drivers/build_all/sensor

> Added. I put it at the end, so the i2c addresses would be used in order. This seems more useful than putting it near the si7006 binding.

Yes it should go at the end, but it doesn't appear that you pushed this change

File: drivers/sensor/si7006/si7006.c
Comment (xyzzy42) [reply drivers/sensor/si7006/si7006.c:7]: You can see the same method being used to change a driver from one device to multiple device here, https://github.com/zephyrproject-rtos/zephyr/pull/48042

It seems the existing policy is that method is just fine and no one had any problems with it.

Comment (MaureenHelm) [reply drivers/sensor/si7006/si7006.c:7]: There isn't a policy one way or the other for the small number of drivers that support more than one compatible. The vast majority of drivers overall use DT_INST macros and diverging from that pattern makes them harder to review.

Comment (xyzzy42) [reply drivers/sensor/si7006/si7006.c:7]: You demanding that I do something differently, which was acceptable when other authors did it, doesn't feel fair and equitable to me.  If this is important, it would be important for other drivers to do it the same way too.  If it's not important, then why the need to demand it from me?

**Comment by teburd** [comment]
Please try to rebase, the failing CI check should go away with a rebase.

**Comment by xyzzy42** [comment]
All failing CI checks are now gone with rebase.

@MaureenHelm, I do not intend to make the change you suggested.  This method is effectively passing implicit parameters to a function via a global variable and I feel that is an anti-pattern to be avoided.  And if anything, the method I have used is more consistent with the existing drivers that support multiple compatible values.

If you feel strongly that what you want is still better, because it makes the drivers with multiple compatibles look more consistent with those with a single compatible, then I offer that demanding I alone change one driver I am writing will not give you this consistency.  New drivers will still use the more common pattern.  And the existing code will still not be written the way you like it.

The existing drivers should be changed to follow the pattern you like too.  And that is clearly out of scope for this PR.

So I propose merging this PR as is, and then as part of the new issue change the existing multi compatible drivers.  That will also give their authors a chance to know about this new design decision.  This way the design of multi-compatible drivers does not need to be decided here in this PR.

**Comment by xyzzy42** [comment]
It has been two months and I can find zero instances of multiple compatible drivers being converted to use the undef DT_DRV_COMPAT pattern.  I guess it's not that important?  Except in this PR, here it is vitally important.  Hard not take that personally.

**Comment by joseppc** [comment]
What's bloking this PR?

**Comment by xyzzy42** [comment]
> Contributor

I think the assignee has no interest.

**Comment by xyzzy42** [comment]
I see in PR #56887 that support for multiple compatible properties was originally done using the repeated undef style that @MaureenHelm prefers, but then changed to use the style that I have used in this PR.  And that PR was approved.

So it seems there are others, like @fabiobaltieri, who do not think the repeated undef system is a good design.

Maybe TSC needs to decide?

**Comment by fabiobaltieri** [comment]
> So it seems there are others, like @fabiobaltieri, who do not think the repeated undef system is a good design.

Hey hi, I just want to clarify that my comment in that PR (https://github.com/zephyrproject-rtos/zephyr/pull/56887#pullrequestreview-1532403532) was more of a general thought than a request for changes, in fact I wrote that comment while approving the PR. Now, I do thin that that undef/define pattern is awful, but if that's what the subsystem uses and the maintainer is happy so be it.

But that's the point: we don't have a clean and well defined approach to handle driver with multiple compatibles and quirks, it'd be nice to come up with something.

For the PR -- apologies for delay, it probably feel off Maureen radar. Is the multi-driver support format the main contention point? If there's no clear guideline, the priority would be the maintainer preference, they are the ones who have to take care of the code after submission after all.

File: drivers/sensor/si7006/si7006.c
Comment (kartben) [inline drivers/sensor/si7006/si7006.c:193]:
```suggestion
	return -ENOTSUP;
```

File: drivers/sensor/si7006/si7006.c
Comment (kartben) [inline drivers/sensor/si7006/si7006.c:111]: Might be worthwhile to convert to switch case?

File: drivers/sensor/si7006/si7006.c
Comment (kartben) [reply drivers/sensor/si7006/si7006.c:193]: > The way the diff chunks appear makes it look like the "unsupported sensor channel" case, but that is above on line 167 and does return `-NOTSUP`.

Exactly what happened yes - sorry :)

File: drivers/sensor/si7006/si7006.c
Comment (fabiobaltieri) [reply drivers/sensor/si7006/si7006.c:7]: Hi @xyzzy42, let me try to unstale the conversation here: coding conventions in the repository are global for few things that are well defined in the coding style, but everything else is then down to the subsystem maintainer and collaborators, and ultimately to the code author. It may feel wrong that the author is the lower priority there but the reality is that once the code is accepted, the maintainer is the one responsible for it in the not so uncommon case that the original author becomes unresponsive for whatever reason, so the code has to feel familiar to them.

This case, handling of mutliple compatibles, does not have a well defined pattern in the coding style. If the maintainer and collaborators have no preferences you can pick your style, but in this case they do, so the right thing to do is to follow their preferred style - nothing personal, no hard feelings, they would have asked it to other authors as well if they were submitting code to the subystem they are responsible for.

File: drivers/sensor/si7006/si7006.c
Comment (xyzzy42) [reply drivers/sensor/si7006/si7006.c:111]: I would have used a case statement if I were writing it.  But it was already written by someone else and with only two possibilities the case statement isn't really any better.  I don't want to makes changes just to follow my style better, that just leads to too much code churn.  Everyone has an opinion.  So I'm just trying to change things to make them simpler or faster or fix bugs or add features, etc.

**Comment by xyzzy42** [comment]
> > So it seems there are others, like @fabiobaltieri, who do not think the repeated undef system is a good design.
> 
> Hey hi, I just want to clarify that my comment in that PR ([#56887 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/56887#pullrequestreview-1532403532)) was more of a general thought than a request for changes, in fact I wrote that comment while approving the PR. Now, I do think that that undef/define pattern is awful, but if that's what the subsystem uses and the maintainer is happy so be it.
> 
> But that's the point: we don't have a clean and well defined approach to handle driver with multiple compatibles and quirks, it'd be nice to come up with something.
> 
> For the PR -- apologies for delay, it probably feel off Maureen radar. Is the multi-driver support format the main contention point? If there's no clear guideline, the priority would be the maintainer preference, they are the ones who have to take care of the code after submission after all.

I've changed to to the undef/define pattern.  All the other sensors drivers use it.

Of course, one might wonder if having different styles mandated in different types of drivers is really the best policy.  What's next, 4 space indents in some code and 8 space in other code?

I don't think the Linux style exactly of using a table can work.  I see no way to avoid the need to use token pasting to construct the macro names generated from the device tree source.  And there's no way to paste tokens other than a macro invocation.  But a more table like macro could probably be written...

```c
DT_FOREACH_STATUS_OKAY_TABLE(MYDEVICE_DEFINE,
        compat_typea, (0x42, true),
        compat_typeb, (0x10, false),
        compat_typec, (0x00, true)
);
```

Linux is limited to one parameter per compat due to the need to fit into an existing `struct of_device_id` or the like.  Zephyr's static device definition style can allow multiple parameters to be passed to each device.

Also notice how with the DT_INST macro version, the macro that creates the device must add an extra argument, `name`, and then construct the object names with a suffix of `##name##_##inst`.   This is because `inst` starts over from 0 on each invocation and can't be used alone to construct a unique object name.  Clearly a table of compats macro would also want to avoid this and allow constructing object names with an extra argument.

**Comment by xyzzy42** [comment]
Pull request was updated and still awaiting review.

**Comment by fabiobaltieri** [comment]
@xyzzy42 could you rebase please?

**Comment by xyzzy42** [comment]
rebased.

Combining the test cases for all sensors into one file causes constant merge conflicts.  Might be better to avoid that by allowing the test to only need to exist in the driver's own files, so there isn't the conflict point of sharing one file with every driver.

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
@MaureenHelm can you take another look? I think that the main point of contention has been addressed.

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by margheritamilani** [comment]
I tried this driver for sht21 sensor and it worked.

**Review by fabiobaltieri — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

**Comment by str4t0m** [comment]
With the conflict https://github.com/zephyrproject-rtos/zephyr/pull/73427 might have a higher chance to go in 3.7.0

**Review by MaureenHelm — DISMISSED** [review]

**Comment by str4t0m** [comment]
Rebased on main to resolve the conflict in `tests/drivers/build_all/sensor/i2c.dtsi`

**Review by jhedberg — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Review by ubieda — APPROVED** [review]

**Comment by aescolar** [comment]
@MaureenHelm please re-approve if you think it is good

**Comment by kartben** [comment]
@aescolar only change was a rebase to solve conflict, it should be good to go

**Pull request merged**
